### ISSUE TITLE
Launch tasks from tasks

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#432]
A commonly requested feature is to emit tasks from other tasks.  There are a variety of dynamic and data-dependent algorithms that are easy to write down in this way.

There are both API and technical considerations to solving this problem.  It's not clear how emitting additional tasks is best accomplished.

### Tornado-like example

Here is an example that looks tornado-y where the parent task depends on the children.

```python
@dask.coroutine
def fib(n):
    if n in (0, 1):
        return n 
    else:
        a, b = yield [fib(i - 1), fib(i - 2)]
        return a + b
```

In this case the worker would start running `fib` in a thread, stop once it got to the `yield` line, send back the two tasks to the scheduler to be run somewhere, and then the computation would restart once those two tasks were finished.  We can do this `yield` trick multiple times for the same function.

This requires us to build a significant amount of logic into the `Worker` (which was previously pretty simple) and also introduces lots of questions about how this plays with cancellation, errors, etc..

### Side effects

Another approach is to assume that the parent task does not depend on its children, but instead is just relying on them for side effects.

```python
def f(i):
    if i > 10:
        emit(f, i / 2)
    if i > 5:
        emit(f, i - 1)
    return i
```

In this case it's not clear who cares about the emitted tasks, so this would only be of use if we were to build the sorts of task-channel machinery necessary for streaming systems.

### Option 3?

    ???

I'm genuinely not certain how people use this sort of thing.  I just know that it's frequently requested.  I encourage people with experience in these sorts of things to speak up about which abstract models are commonly found in the wild and what their strengths and drawbacks are.

My guess is that people prefer the tornado-like solution.  This is non-trivial but feasible with a fair amount of effort.  Just to set expectations, this is nowhere on my immediate roadmap.

**Comment by spitz-dan-l** [comment]
Definitely an important issue, thanks for writing it up.

I worry that if we go with the tornado-like route, a lot of refactoring of existing code using dask distributed would be in order. I find that a common use case I am running into is that I'll have some non-task code, which uses an executor and tasks, and waits for them. I'll then want to run this "non-task" as a task: multiple times or for multiple inputs. With the tornado-like approach, this would *always* involve refactoring the non-task code to use `yield`, and I'd constantly be wondering if I should implement things using `yield` up-front or as top-level non-task code. I _think_ that this is part of what the new `async` syntax in Python 3.5 is meant to solve- just use `async` everywhere and don't worry about it. (And it might be that it is just the inevitable solution in cases like this, I'm not sure.)

I have used a pretty hacky workaround to the issue up til now. It's basically just, "use multiple schedulers". In my use case, there is no unbounded branching of tasks (as in the fibonacci example); there are just two levels- Level 1 always calls out to, and waits for, Level 2. And top level always invokes Level 1. So I have a scheduler for Level 1 and a scheduler for Level 2, each with their own workers. Works fine.

One nice thing to have would be some support for this pattern, without need for multiple schedulers. I think we'd need the concept of worker pools, and the ability of Workers reason about "which worker pool am I in?" so they can error if you try to submit a task in your own pool. You could just write:

```
def f1(x):
    e = Executor(THE_EXECUTOR_URI)
    fs = e.map(f2, list(range(x)), worker_pool='level_2_pool') #works if not already in level 2 pool
    return sum(f.result() for f in fs())

def f2(i):
    return i**2

e = Executor(THE_EXECUTOR_URI)
fs = e.map(f1, list(range(10)), worker_pool='level_1_pool') #works if not already in level 1 pool
print(list(f.result() for f in fs))
```

The upside is it *might* be less work to implement (I'm talking outta my butt here) and would not require refactoring all code to use `yield`.

In any case, keep up the excellent work on Dask! It rocks!

**Comment by mrocklin** [comment]
@spitz-dan-l something like what you're looking for has been implemented in https://github.com/dask/distributed/pull/471

**Comment by spitz-dan-l** [comment]
@mrocklin Looks great, thanks!

**Comment by mrocklin** [comment]
Fixed in #471
